### PR TITLE
Refactor bandstructure routines and json results

### PR DESCRIPTION
**Pull request by mfherbst**
- [x] Finish tests
- [x] Inline functions in bzmesh.jl

@azadoks Was this what you had in mind as a minimal output for the band data ? I will make an HDF5 version as well, that allows to save the bloch waves as well.

Supplying arbitrary k-point coordinates is now also easy.

### CODE REVIEW

**Comment by azadoks** [comment]
It would be nice to have:

- The k-points in fractional or Cartesian coordinates (or both)
- Labels for the high-symmetry k-points (if DFTK has generated the high-symmetry path)
- The direct or reciprocal lattice (so that you can convert the k-points to fractional or Cartesian coordinates if you need to)
- The k-weights (with a sum normalized to 1.0, 2.0, or the number of k-points or twice the number of k-points; it's a choice to make but be consistent; QE normalizes the sum to 2.0)
- The relevant symmetry operators (so you can recover the full BZ from the eigenvalues and k-points in the IBZ wedge)
- The eigenvalues for each band, k-point, spin channel
- The occupations for each band, k-point, spin channel
- The Fermi level
- The number of electrons (if you want to re-compute the Fermi level)

This way, computing the Gaussian (k-weights, eigenvalues) and tetrahedral integrated (k-points, symmetries, eigenvalues) DOS should both be straightforward.

Making nice plots for the bands on a path (k-points, k-point labels, eigenvalues, Fermi level) should also be straightforward.

Then, for projected DOS and bands, one would need the Bloch waves too.

**Comment by antoine-levitt** [comment]
I don't know what this is for but it looks very close to the dftk - Wannier interface, so maybe use that?

**Comment by mfherbst** [comment]
Thanks for the comments @azadoks. I'm not quite sure about some of them in the context of band structure data, for example

> The k-weights (with a sum normalized to 1.0, 2.0, or the number of k-points or twice the number of k-points; it's a choice to make but be consistent; QE normalizes the sum to 2.0)

Does this make sense here? I mean the k-Path is not designed to give a useful integration, in particular it is not necessarily a useful IBZ wedge. One could probably use the symmetries to unfold it, but is that worth it ?

But the related question (related also to what Antoine wrote): Is it even reasonable to have a completely separate `compute_bands` routine as we have now, which is intended just for bands? Maybe one should instead have just a unified way to recompute the eigenvalues given a converged SCF at "k points". This could serve both for doing this at a larger MP-grid / BZ wedge (where it would make sense to do the integration stuff on top) and at arbitrary k-Points (where it might not make sense). That indeed should be useful for wannier-related things when one wants to use more k-Points for the interpolation than for the SCF and for plotting bands as well.

**Comment by azadoks** [comment]
On a high-symmetry line, indeed the weights do not make sense. However, consistent output is 
> Maybe one should instead have just a unified way to recompute the eigenvalues given a converged SCF at "k points".

I do think this is the way to go, maybe with a dispatch that generates the k-path from a minimal spacing or line density.
In the end, this would be similar to QE's `nscf` calculation; I'm not really sure why they have two modes: `nscf` and `bands`. 

> I mean the k-Path is not designed to give a useful integration, in particular it is not necessarily a useful IBZ wedge. One could probably use the symmetries to unfold it, but is that worth it ?

For explicit lists of k-points or when DFTK has used Brillouin to generate a k-path (still an explicit list), then the weights don't make sense and symmetries aren't used in the calculation itself, so it doesn't necessarily make sense to write them out.
If the calculation is on a Monkhorst-Pack mesh and done with symmetries, then it does make sense to write out both the symmetries and the weights, IMO.
Having conditional output can sometimes be a bit annoying, which is why I'd still support always writing at least _something_, but that's personal preference.

**Comment by mfherbst** [comment]
@azadoks I took most of your comments into consideration the `save_bands` with json output now has pretty much all of this and extending it towards HDF5, including bloch waves, should not be too hard once this PR is in.

@antoine-levitt The main changes of the PR on the structural DFTK side is a harmonisation of the return values of `self_consistent_field` and `compute_bands`. As a result `compute_bands` can now also be used to top-up the k-point grid without self-consistency. I'm thinking about Wannier, DOS and similar computations where that could be useful. To make our k-point treatment more homogeneous between the two use cases (explicit k-points for bands and MP grid) I introduced some data structures, which should also be useful to optimise a few things (e.g. `symmetries_preserving_kgrid` we can improve for that case).

**Comment by antoine-levitt** [comment]
Looks reasonable. Not a big fan of kgrid meaning either an object or a vector though... Maybe leave it alone until we get around to splitting the kpoint mesh structure?

**Comment by mfherbst** [comment]
> kgrid meaning either an object or a vector though

Me neither. I could not figure out a good way to distinguish though.

Shall we leave it as is for now and unify the language along with the k-mesh / q-mesh things or you rather want to settle on something now and work it in consistently with this PR ?

**Comment by antoine-levitt** [comment]
Nvm I misunderstood, sorry, let me take a closer look

**Comment by antoine-levitt** [comment]
OK sorry, this makes sense as it is, LGTM

File: examples/collinear_magnetism.jl [hunk 1]
```diff
@@ -101,11 +101,14 @@ idown = iup + length(scfres.basis.kpoints) ÷ 2
...
+plot_dos(bands_666)
+# Note that if same k-grid as SCF should be employed, a simple `plot_dos(scfres)`
+# is sufficient.
 
 # Similarly the band structure shows clear differences between both spin components.
 using Unitful
 using UnitfulAtomic
-plot_bandstructure(scfres; kline_density=6)
+bands_kpath = compute_bands(scfres; kline_density=6)
+plot_bandstructure(bands_kpath)
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply examples/collinear_magnetism.jl:114]: My long-term plan is to integrate with plot recipes where you essentially can directly plot the band data in a meaningful way and interact with it (e.g. overlay two plots etc, change axes and labels etc.). So in the long run a simple standard plot would become `plot(compute_bands(scfres; kline_density=6))`, which is essentially the same we have now.

[the remark this answers is not in the document]
Comment (antoine-levitt) [reply examples/collinear_magnetism.jl:114]: I definitely agree it's nice to split, but this is needlessly breaking, so we might want to keep it as is a little longer (or deprecate it, like you've done with the other method)

[the remark this answers is not in the document]
Comment (mfherbst) [reply examples/collinear_magnetism.jl:114]: It is only deprecated. The original `plot_bandstructure(scfres)` should still work.

[the remark this answers is not in the document]
Comment (antoine-levitt) [reply examples/collinear_magnetism.jl:114]: OK sorry I misread then

**Comment by azadoks** [comment]
> Me neither. I could not figure out a good way to distinguish though.

This has always bothered me as well. Maybe what's done with the FFT grid could transfer?
For that, we have `fft_size` for the shape/size and `Gvectors` for the array of vectors/points. Could be something like `kgrid_size`, `k_grid_size`, `mp_size` (for Monkhorst-Pack, where this makes sense) for the shape/size and `kvectors` or `kpoints` for the array of vectors/points.

**Comment by mfherbst** [comment]
> Could be something like kgrid_size, k_grid_size, mp_size

That might work. The only think is that some of the names are maybe a bit clunky. We could settle on `mpsize` or `ksize` for the bare `(6, 6, 6)` thing, `kgrid` for the object respresenting the k-point mesh. This is a pretty heavy breaking change though as `kgrid` is used pretty much in very DFTK script right now, so we should make sure this is really what we want.

> for the array of vectors/points.

I think for that we already consistently use `kcoords` or `kpoints`

**Comment by azadoks** [comment]
> This is a pretty heavy breaking change though as kgrid is used pretty much in very DFTK script right now, so we should make sure this is really what we want.

Ah then maybe the juice isn't worth the squeeze. This could always be changed later if deemed necessary

**Pull request merged**